examples/demo.ipynb

```diff
@@ -0,0 +1,625 @@
+{
+ "cells": [
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "Copyright (c) 2024 Graphcore Ltd. All rights reserved.\n",
+    "\n",
+    "# u-μP using the `unit_scaling` library\n",
+    "\n",
+    "This notebook walks through the definition and training of a byte-level language model, comparing the [Unit-Scaled Maximal Update Parametrization](https://openreview.net/forum?id=44NKKzz1n5) \"u-μP\" against Standard Parametrization \"SP\" on learning-rate transfer across width.\n",
+    "\n",
+    "We show:\n",
+    " - How to define a u-μP model.\n",
+    " - The forward and backward-pass scaling properties of a u-μP transformer decoder layer, versus an SP layer.\n",
+    " - Learning rate transfer when training a u-μP model."
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "## Preamble"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 1,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "import re\n",
+    "from typing import *\n",
+    "\n",
+    "import datasets\n",
+    "import einops\n",
+    "import matplotlib.pyplot as plt\n",
+    "import pandas as pd\n",
+    "import seaborn as sns\n",
+    "from tqdm import tqdm\n",
+    "\n",
+    "import torch\n",
+    "from torch import nn, Tensor, tensor\n",
+    "import torch.nn.functional as F\n",
+    "\n",
+    "import unit_scaling as uu\n",
+    "import unit_scaling.functional as U\n",
+    "\n",
+    "# Config & helpers\n",
+    "torch.backends.cuda.matmul.allow_tf32 = True\n",
+    "torch.backends.cudnn.allow_tf32 = True\n",
+    "def show_layer_stats(layer: nn.Module, input_shape: Tuple[int, ...]) -> None:\n",
```

Comment:
Demo notebook looks good! Only thing that broke my attention was searching for `show_layer_stats`. Was  easy to miss when colocated with imports. My preference would be to have it defined next to first use, or at least be in its own cell.